lib/client/session.go

```diff
@@ -593,41 +598,76 @@ func (ns *NodeSession) pipeInOut(shell io.ReadWriteCloser) {
 			log.Errorf(err.Error())
 		}
 	}()
-	// copy from the local input to the remote shell:
-	go func() {
-		defer ns.closer.Close()
-		buf := make([]byte, 128)
-
-		stdin := ns.terminal.Stdin()
-		if ns.terminal.IsAttached() && ns.enableEscapeSequences {
-			stdin = escape.NewReader(stdin, ns.terminal.Stderr(), func(err error) {
-				switch err {
-				case escape.ErrDisconnect:
-					fmt.Fprintf(ns.terminal.Stderr(), "\r\n%v\r\n", err)
-				case escape.ErrTooMuchBufferedData:
-					fmt.Fprintf(ns.terminal.Stderr(), "\r\nerror: %v\r\nremote peer may be unreachable, check your connectivity\r\n", trace.Wrap(err))
-				default:
-					fmt.Fprintf(ns.terminal.Stderr(), "\r\nerror: %v\r\n", trace.Wrap(err))
+
+	switch mode {
+	case types.SessionObserverMode:
+		fallthrough
+	case types.SessionModeratorMode:
+		go func() {
+			defer ns.closer.Close()
+
+			for {
```

Comment:
I'm not sure what you mean. `lib/client/session.go` implements the SSH client. This change was made since we're supporting moderated sessions for SSH sessions also (original scope was expanded during a demo).

Reply:
Sorry, I meant in the kube sessions.

You have an almost identical loop in lib/client/kubesession.go:

- allocate 1-byte slice
- read from stdin
- check for EOF
- compare byte to ctrl-c or ctrl-t

It seems like there's enough in common that you could combine these two into a shared function rather than duplicate the code. Worth investigating at least.